packages/flutter/lib/src/rendering/sliver.dart

```diff
@@ -1742,46 +1440,599 @@ class RenderViewport2 extends RenderBox with ContainerRenderObjectMixin<RenderSl
+    @required ViewportOffset offset,
+    double anchor: 0.0,
+    List<RenderSliver> children,
+    RenderSliver center,
+  }) : _anchor = anchor,
+       _center = center,
+       super(axisDirection: axisDirection, offset: offset) {
+    assert(anchor != null);
+    assert(anchor >= 0.0 && anchor <= 1.0);
+    addAll(children);
+    if (center == null && firstChild != null)
+      _center = firstChild;
+  }
+
+  @override
+  void setupParentData(RenderObject child) {
+    if (child.parentData is! SliverPhysicalContainerParentData)
+      child.parentData = new SliverPhysicalContainerParentData();
+  }
+
+  double get anchor => _anchor;
+  double _anchor;
+  set anchor(double value) {
+    assert(value != null);
+    assert(value >= 0.0 && value <= 1.0);
+    if (value == _anchor)
+      return;
+    _anchor = value;
+    markNeedsLayout();
+  }
+
+  RenderSliver get center => _center;
+  RenderSliver _center;
+  set center(RenderSliver value) {
+    if (value == _center)
+      return;
+    _center = value;
+    markNeedsLayout();
+  }
+
+  @override
+  bool get sizedByParent => true;
+
+  @override
+  void performResize() {
+    assert(constraints.hasBoundedHeight && constraints.hasBoundedWidth);
+    size = constraints.biggest;
+    switch (axis) {
+      case Axis.vertical:
+        offset.applyViewportDimension(size.height);
+        break;
+      case Axis.horizontal:
+        offset.applyViewportDimension(size.width);
+        break;
+    }
+  }
+
+  // Out-of-band data computed during layout.
+  double _minScrollExtent;
+  double _maxScrollExtent;
+  bool _hasVisualOverflow = false;
+
+  @override
+  void performLayout() {
     if (center == null) {
       assert(firstChild == null);
-      return result;
+      _minScrollExtent = 0.0;
+      _maxScrollExtent = 0.0;
+      _hasVisualOverflow = false;
+      offset.applyContentDimensions(0.0, 0.0);
+      return;
     }
+
     assert(center.parent == this);
-    assert(firstChild != null);
+    assert(constraints.hasBoundedHeight && constraints.hasBoundedWidth);
+    double extent;
+    double crossExtent;
+    switch (axis) {
+      case Axis.vertical:
+        extent = size.height;
+        crossExtent = size.width;
+        break;
+      case Axis.horizontal:
+        extent = size.width;
+        crossExtent = size.height;
+        break;
+    }
+
+    final double centerOffsetAdjustment = center.centerOffsetAdjustment;
+
+    double correction;
+    do {
+      assert(offset.pixels != null);
+      correction = _attemptLayout(extent, crossExtent, offset.pixels + centerOffsetAdjustment);
+      if (correction != 0.0) {
+        offset.correctBy(correction);
+      } else {
+        // when updating this, also update rereportDimensions
+        if (offset.applyContentDimensions(
+              math.min(0.0, _minScrollExtent + extent * anchor),
+              math.max(0.0, _maxScrollExtent - extent * (1.0 - anchor)),
+           ))
+          break;
+      }
+    } while (true);
```

Comment:
We should probably count down and throw an exception if we spin this loop more than, for example, 10 times.